File: cs/BowlingGame/Solved/Game.cs
using NUnit.Framework;

namespace BowlingGame.Solved
{
    //Перевод на C# оригинального решения от дядюшки Боба
    public class Game
    {
        private readonly int[] rolls = new int[21];
        private int currentRoll;

        public void Roll(int pins)
        {
            rolls[currentRoll++] = pins;
        }

        public int GetScore()
        {
            int score = 0;
            int frameIndex = 0;
            for (int frame = 0; frame < 10; frame++)
            {
                if (IsStrike(frameIndex))
                {
                    score += 10 + GetStrikeBonus(frameIndex);
                    frameIndex++;
                }
                else if (IsSpare(frameIndex))
                {
                    score += 10 + GetSpareBonus(frameIndex);
                    frameIndex += 2;
                }
                else
                {
                    score += GetSumOfBallsInFrame(frameIndex);
                    frameIndex += 2;
                }
            }
            return score;
        }

        private bool IsStrike(int frameIndex)
        {
            return rolls[frameIndex] == 10;
        }

        private int GetSumOfBallsInFrame(int frameIndex)
        {
            return rolls[frameIndex] + rolls[frameIndex + 1];
        }

        private int GetSpareBonus(int frameIndex)
        {
            return rolls[frameIndex + 2];
        }

        private int GetStrikeBonus(int frameIndex)
        {
            return rolls[frameIndex + 1] + rolls[frameIndex + 2];
        }

        private bool IsSpare(int frameIndex)
        {
            return rolls[frameIndex] + rolls[frameIndex + 1] == 10;
        }
    }
    
    [TestFixture]
	public class Game_Tests
    {
        private Game game;

        [SetUp]
        public void SetUp()
        {
            game = new Game();
        }

        [Test]
        public void TestGutterGame()
        {
            RollMany(20, 0);
            Assert.AreEqual(0, game.GetScore());
        }

        [Test]
        public void TestAllOnes()
        {
            RollMany(20,1);
            Assert.AreEqual(20, game.GetScore());
        }

        [Test]
        public void TestOneSpare()
        {
            RollSpare();
            game.Roll(3);
            RollMany(17,0);
            Assert.AreEqual(16,game.GetScore());
        }

        [Test]
        public void TestOneStrike()
        {
            RollStrike();
            game.Roll(3);
            game.Roll(4);
            RollMany(16, 0);
            Assert.AreEqual(24, game.GetScore());
        }

        [Test]
        public void TestPerfectGame()
        {
            RollMany(12,10);
            Assert.AreEqual(300, game.GetScore());
        }

        private void RollMany(int n, int pins)
        {
            for (int i = 0; i < n; i++)
                game.Roll(pins);
        }

        private void RollStrike()
        {
            game.Roll(10);
        }

        private void RollSpare()
        {
            game.Roll(5);
            game.Roll(5);
        }
    }
}

File: js/src/bowlingGame/.solved/game.js
import "chai/register-assert";

//Перевод на JS оригинального решения от дядюшки Боба
class Game {
    constructor() {
        this.rolls = [];
        this.currentRoll = 0;
    }

    roll(pins) {
        this.rolls[this.currentRoll++] = pins;
    }

    getScore() {
        let score = 0;
        let frameIndex = 0;
        for (let frame = 0; frame < 10; frame++) {
            if (this.isStrike(frameIndex)) {
                score += 10 + this.getStrikeBonus(frameIndex);
                frameIndex++;
            }
            else if (this.isSpare(frameIndex)) {
                score += 10 + this.getSpareBonus(frameIndex);
                frameIndex += 2;
            }
            else {
                score += this.getSumOfBallsInFrame(frameIndex);
                frameIndex += 2;
            }
        }
        return score;
    }

    isStrike(frameIndex) {
        return this.rolls[frameIndex] === 10;
    }

    getSumOfBallsInFrame(frameIndex) {
        return this.rolls[frameIndex] + this.rolls[frameIndex + 1];
    }

    getSpareBonus(frameIndex) {
        return this.rolls[frameIndex + 2];
    }

    getStrikeBonus(frameIndex) {
        return this.rolls[frameIndex + 1] + this.rolls[frameIndex + 2];
    }

    isSpare(frameIndex) {
        return this.rolls[frameIndex] + this.rolls[frameIndex + 1] === 10;
    }
}

describe("Game tests", () => {
    let game;

    beforeEach(() => {
        game = new Game();
    });

    it("test gutter game", () => {
        rollMany(20, 0);
        assert.equal(game.getScore(), 0);
    });

    it("test all ones", () => {
        rollMany(20,1);
        assert.equal(game.getScore(), 20);
    });

    it("test one spare", () => {
        rollSpare();
        game.roll(3);
        rollMany(17,0);
        assert.equal(game.getScore(), 16);
    });

    it("test one strike", () => {
        rollStrike();
        game.roll(3);
        game.roll(4);
        rollMany(16, 0);
        assert.equal(game.getScore(), 24);
    });

    it("test perfect game", () => {
        rollMany(12,10);
        assert.equal(game.getScore(), 300);
    });

    function rollMany(n, pins) {
        for (let i = 0; i < n; i++) {
            game.roll(pins);
        }
    }

    function rollStrike() {
        game.roll(10);
    }

    function rollSpare() {
        game.roll(5);
        game.roll(5);
    }
});


